File: internal/webui/service/evals.go
package service

import (
	"context"
	"time"

	"github.com/tysonthomas9/loomcli/internal/domain"
)

// EvalAdminService defines the webui-facing session evaluation administration
// operations. Handlers own HTTP parsing; implementations own store/evals
// orchestration and return typed service errors.
type EvalAdminService interface {
	GetRollup(ctx context.Context, wsID string, opts EvalRollupOptions) (*EvalRollupData, error)
	GetSessionEvalState(ctx context.Context, wsID, sessionID string) (*SessionEvalState, error)
	RejudgeSession(ctx context.Context, wsID, sessionID string) (*EvalRejudgeResult, error)
	GetCron(ctx context.Context, wsID string) (*EvalCronState, error)
	SetCronEnabled(ctx context.Context, wsID string, enabled bool) (*EvalCronState, error)
}

type EvalRollupOptions struct {
	Since time.Time
	Until time.Time
}

type EvalScoreAverages struct {
	OutcomeSuccess       float64 `json:"outcome_success"`
	InstructionAdherence float64 `json:"instruction_adherence"`
	Efficiency           float64 `json:"efficiency"`
	ToolUseQuality       float64 `json:"tool_use_quality"`
}

type EvalScoreBucket struct {
	BucketStart time.Time         `json:"bucket_start"`
	EvalCount   int               `json:"eval_count"`
	Averages    EvalScoreAverages `json:"averages"`
}

type EvalCountByTag struct {
	Tag   string `json:"tag"`
	Count int    `json:"count"`
}

type EvalInsight struct {
	Text      string    `json:"text"`
	SessionID string    `json:"session_id"`
	EvalID    string    `json:"eval_id"`
	CreatedAt time.Time `json:"created_at"`
}

type EvalInsightCategories struct {
	Harness []EvalInsight `json:"harness"`
	Linter  []EvalInsight `json:"linter"`
	Prompt  []EvalInsight `json:"prompt"`
	Skill   []EvalInsight `json:"skill"`
}

type EvalFailureClassCount struct {
	ErrorClass string `json:"error_class"`
	Count      int    `json:"count"`
}

type EvalJudgePromptVersionCount struct {
	Version string `json:"version"`
	Count   int    `json:"count"`
}

type EvalRollupData struct {
	Since               time.Time                     `json:"since"`
	Until               time.Time                     `json:"until"`
	EvalCount           int                           `json:"eval_count"`
	ScoreAverages       EvalScoreAverages             `json:"score_averages"`
	ScoreBuckets        []EvalScoreBucket             `json:"score_buckets"`
	TagFrequencies      []EvalCountByTag              `json:"tag_frequencies"`
	Insights            EvalInsightCategories         `json:"insights"`
	FailureClasses      []EvalFailureClassCount       `json:"failure_classes"`
	JudgePromptVersions []EvalJudgePromptVersionCount `json:"judge_prompt_versions"`
}

type SessionEvalState struct {
	EvalStatus        string              `json:"eval_status"`
	EvalErrorClass    *string             `json:"eval_error_class"`
	EvalPromptVersion *string             `json:"eval_prompt_version"`
	EvalRequested     bool                `json:"eval_requested"`
	Eval              *domain.SessionEval `json:"eval"`
}

type EvalRejudgeResult struct {
	Requested      bool `json:"requested"`
	BindingEnabled bool `json:"binding_enabled"`
}

type EvalCronState struct {
	Provisioned bool    `json:"provisioned"`
	Enabled     bool    `json:"enabled"`
	Schedule    *string `json:"schedule"`
}

File: internal/webui/frontend/src/types/agent/evals.ts
export type EvalScoreKey =
  | "outcome_success"
  | "instruction_adherence"
  | "efficiency"
  | "tool_use_quality";

/** Dimension-keyed scores. New rubrics may add dimensions without a UI release. */
export type EvalScores = Record<string, number>;

export interface EvalScoreAverages {
  outcome_success: number;
  instruction_adherence: number;
  efficiency: number;
  tool_use_quality: number;
}

export interface EvalScoreBucket {
  bucket_start: string;
  eval_count: number;
  averages: EvalScoreAverages;
}

export interface EvalTagFrequency {
  tag: string;
  count: number;
}

export interface EvalInsight {
  text: string;
  session_id: string;
  eval_id: string;
  created_at: string;
}

export interface EvalInsightCategories {
  harness: EvalInsight[];
  linter: EvalInsight[];
  prompt: EvalInsight[];
  skill: EvalInsight[];
}

export interface EvalFailureClassCount {
  error_class: string;
  count: number;
}

export interface EvalJudgePromptVersionCount {
  version: string;
  count: number;
}

export interface EvalRollupData {
  since: string;
  until: string;
  eval_count: number;
  score_averages: EvalScoreAverages;
  score_buckets: EvalScoreBucket[];
  tag_frequencies: EvalTagFrequency[];
  insights: EvalInsightCategories;
  failure_classes: EvalFailureClassCount[];
  judge_prompt_versions: EvalJudgePromptVersionCount[];
}

export type SessionEvalStatus = "done" | "failed" | "none";

export interface SessionEvalScoreRationales {
  outcome_success: string;
  instruction_adherence: string;
  efficiency: string;
  tool_use_quality: string;
}

export interface SessionEvalCost {
  input_tokens: number;
  output_tokens: number;
  total_tokens: number;
}

export interface SessionEvalRecord {
  eval_id: string;
  session_id: string;
  task_id?: string;
  agent_id: string;
  workspace_key: string;
  scores: EvalScores;
  score_rationales: SessionEvalScoreRationales;
  error_taxonomy_tags: string[];
  improvement_categories: {
    harness: string[];
    linter: string[];
    prompt: string[];
    skill: string[];
  };
  judge_summary: string;
  judge_model: string;
  judge_prompt_version: string;
  /** Judge AgentSession that produced this exact eval record, when available. */
  judge_session_id?: string;
  eval_cost: SessionEvalCost;
  session_started_at: string;
  session_ended_at: string;
  created_at: string;
  updated_at: string;
}

export interface SessionEvalState {
  eval_status: SessionEvalStatus;
  eval_error_class?: string | null;
  eval_prompt_version?: string | null;
  eval_requested: boolean;
  eval: SessionEvalRecord | null;
}

export interface EvalCronState {
  provisioned: boolean;
  enabled: boolean;
  schedule?: string | null;
}

export interface EvalRejudgeResult {
  requested: boolean;
  binding_enabled: boolean;
}
